### PR TITLE
Add UI changes for kube exec functionality.

### PR DESCRIPTION
**Pull request by AntonAM**
This PR adds UI for the Kube Exec functionality (backend part was added in #41144).
Mainly it changes appearance of connect dialog for the kube resource and adds new type of document for the console - `DocumentKubeExec`, which is very similar to the `DocumentSsh`. Also we allow new option for the terminal `ConvertEol` that correctly displays output for non-interactive shell, because we get just `\n` instead of `\r\n` newlines in that mode.

UI changes for the Connect dialog:
kubeexec_connect_dialog

Demo in action:

https://github.com/gravitational/teleport/assets/964862/6ecce7c9-00b4-40fc-a6e3-679a9f98157c.mov

Changelog: Add ability to exec into a pod on a Kubernetes cluster using Web UI.

### CODE REVIEW

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.test.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      addEventListener: jest.fn(),
+      removeEventListener: jest.fn(),
+      dispatchEvent: jest.fn(),
+    })),
+  });
+
+  const { container } = render(
+    <DocumentKubeExecWrapper ctx={ctx} consoleCtx={consoleCtx} doc={baseDoc} />
+  );
+  expect(container.firstChild).toMatchSnapshot();
```
Comment (ravicious) [inline web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.test.tsx:49]: I don't view this strictly as a requirement since I don't think you usually work on frontend stuff, but if you have time then I'd timebox this to like 30–60 minutes and see how far you can get with the idea below, it'd help us out in the long term! If not then I guess the snapshot test is fine.

---

I understand that this was mostly copied from `DocumentSsh`, but it'd be nice to add a regular test instead of a snapshot test, as we generally want to reduce the number of snapshot tests.

It could boil down to duplicating the setup from the story (so that the test does not depend on the story) and then performing a simple check. It could be just writing to the tty and then checking if that gets rendered on the screen.

If all else fails, the check could end up as `data-testid="terminal"` added to the div rendered by `Terminal`:

https://github.com/gravitational/teleport/blob/43d050b82158773fd20d23223308d80be5160e3d/web/packages/teleport/src/Console/DocumentSsh/Terminal/Terminal.tsx#L169-L177

…and then checking if the div got rendered on the screen:

https://github.com/gravitational/teleport/blob/43d050b82158773fd20d23223308d80be5160e3d/web/packages/teleport/src/components/ToolTipNoPermBadge/ToolTipNoPermBadge.test.tsx#L37

(Though using `data-testid` should be the last resort.)

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+  const noSidDoc = {
+    ...baseDoc,
+    sid: '',
+  };
+
+  return (
+    <DocumentKubeExecWrapper ctx={ctx} consoleCtx={consoleCtx} doc={noSidDoc} />
+  );
+};
```
Comment (ravicious) [inline web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx:73]: In storybook, these three all render as the same to me, I'm not sure if that's what's supposed to happen.

I wanted to compare this against `DocumentSsh` stories, but it looks like stories for that component have been broken for some time.

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+export type Props = {
+  ctx: TeleportContext;
+  consoleCtx: ConsoleCtx;
+  doc: stores.DocumentKubeExec;
+};
```
Comment (ravicious) [inline web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx:79]: This is not used and can be removed I suppose.

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  kubeCluster: 'kubeCluster1',
+  kubeNamespace: 'namespace1',
+  pod: 'pod1',
+  container: '',
+  id: 3,
+  url: 'fd',
+  created: new Date(),
+  command: '/bin/bash',
+  isInteractive: true,
+} as const;
```
Comment (ravicious) [inline web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx:118]: That's one of the problems with exporting stuff out of stories and then importing it in tests. On a conceputal level, I think it's not what one's supposed to (though that's the pattern I think we've been following in a few places).

But a practical problem with it is that Storybook expects each exported member to be a React component, so we end up with a bunch of items that look like stories, but they just don't work.

exported-stories

File: web/packages/teleport/src/Kubes/ConnectDialog/ConnectDialog.tsx [hunk 5]
```diff
@@ -99,6 +133,64 @@ function ConnectDialog(props: Props) {
...
+                  </Box>
+                  <ToolTipInfo>
+                    You can start an interactive shell and have a bidirectional
+                    communication with the target pod, or you can run one-off
+                    command and see its output.
+                  </ToolTipInfo>
+                </Toggle>
+                <ButtonPrimary onClick={startKubeExecSession}>
+                  Run Command
+                </ButtonPrimary>
```
Comment (ravicious) [inline web/packages/teleport/src/Kubes/ConnectDialog/ConnectDialog.tsx:189]: If we wrap the form into `<form>`, add `onSubmit` on it and change the button's type to `submit` and then add autofocus on the first field, then it'll be possible to navigate the form using keyboard and submit it with Enter.

<details>
<summary>Patch</summary>

Copy then `pbpaste | git apply`

```diff
diff --git a/web/packages/teleport/src/Kubes/ConnectDialog/ConnectDialog.tsx b/web/packages/teleport/src/Kubes/ConnectDialog/ConnectDialog.tsx
index eb06d90ba8..ed7e858f08 100644
--- a/web/packages/teleport/src/Kubes/ConnectDialog/ConnectDialog.tsx
+++ b/web/packages/teleport/src/Kubes/ConnectDialog/ConnectDialog.tsx
@@ -56,7 +56,8 @@ function ConnectDialog(props: Props) {
   const [execCommand, setExecCommand] = useState('');
   const [execInteractive, setExecInteractive] = useState(true);
 
-  const startKubeExecSession = () => {
+  const startKubeExecSession = event => {
+    event.preventDefault();
     const splitPath = execPath.split('/');
 
     const url = cfg.getKubeExecConnectRoute(
@@ -136,58 +137,59 @@ function ConnectDialog(props: Props) {
         <Validation>
           {() => (
             <Box borderTop={1} mb={4} mt={4}>
-              <Text mt={3} bold>
-                Or exec into a pod on this Kubernetes cluster
-              </Text>
-              <Flex gap={3}>
-                <FieldInput
-                  value={execPath}
-                  placeholder="namespace/pod"
-                  label="Pod to exec into"
-                  width="50%"
-                  onChange={e => setExecPath(e.target.value.trim())}
-                  toolTipContent={
-                    <Text>
-                      Specify namespace and pod you want to exec into.
-                      Optionally you can also specify container by adding it at
-                      the end: '/namespace/pod/container'
-                    </Text>
-                  }
-                />
-                <FieldInput
-                  rule={requiredField('Command to execute is required')}
-                  value={execCommand}
-                  placeholder="/bin/bash"
-                  label="Command to execute"
-                  width="50%"
-                  onChange={e => setExecCommand(e.target.value)}
-                  toolTipContent={
-                    <Text>
-                      The command that will be executed inside the target pod.
-                    </Text>
-                  }
-                />
-              </Flex>
-              <Flex justifyContent="space-between" gap={3}>
-                <Toggle
-                  isToggled={execInteractive}
-                  onToggle={() => {
-                    setExecInteractive(b => !b);
-                  }}
-                >
-                  <Box ml={2} mr={1}>
-                    Interactive shell
-                  </Box>
-                  <ToolTipInfo>
-                    You can start an interactive shell and have a bidirectional
-                    communication with the target pod, or you can run one-off
-                    command and see its output.
-                  </ToolTipInfo>
-                </Toggle>
-                <ButtonPrimary onClick={startKubeExecSession}>
-                  Run Command
-                </ButtonPrimary>
-              </Flex>
+              <form onSubmit={startKubeExecSession}>
+                <Text mt={3} bold>
+                  Or exec into a pod on this Kubernetes cluster
+                </Text>
+                <Flex gap={3}>
+                  <FieldInput
+                    value={execPath}
+                    placeholder="namespace/pod"
+                    autoFocus
+                    label="Pod to exec into"
+                    width="50%"
+                    onChange={e => setExecPath(e.target.value.trim())}
+                    toolTipContent={
+                      <Text>
+                        Specify namespace and pod you want to exec into.
+                        Optionally you can also specify container by adding it
+                        at the end: '/namespace/pod/container'
+                      </Text>
+                    }
+                  />
+                  <FieldInput
+                    rule={requiredField('Command to execute is required')}
+                    value={execCommand}
+                    placeholder="/bin/bash"
+                    label="Command to execute"
+                    width="50%"
+                    onChange={e => setExecCommand(e.target.value)}
+                    toolTipContent={
+                      <Text>
+                        The command that will be executed inside the target pod.
+                      </Text>
+                    }
+                  />
+                </Flex>
+                <Flex justifyContent="space-between" gap={3}>
+                  <Toggle
+                    isToggled={execInteractive}
+                    onToggle={() => {
+                      setExecInteractive(b => !b);
+                    }}
+                  >
+                    <Box ml={2} mr={1}>
+                      Interactive shell
+                    </Box>
+                    <ToolTipInfo>
+                      You can start an interactive shell and have a
+                      bidirectional communication with the target pod, or you
+                      can run one-off command and see its output.
+                    </ToolTipInfo>
+                  </Toggle>
+                  <ButtonPrimary type="submit">Run Command</ButtonPrimary>
+                </Flex>
+              </form>
             </Box>
           )}
         </Validation>

```
</details>

**Review by ravicious — COMMENTED** [review]
I'll continue the review on Monday.

File: web/packages/teleport/src/config.ts [hunk 6]
```diff
@@ -569,6 +574,15 @@ const cfg = {
     });
   },
 
+  getKubeExecConnectRoute(params: UrlKubeExecParams, query: UrlKubeExecQuery) {
+    if (!params.container) {
+      params.container = undefined;
+    }
```
Comment (ravicious) [inline web/packages/teleport/src/config.ts:580]: Why is this needed?

**Review by ravicious — COMMENTED** [review]
The biggest problem for me with this PR is that it adds the ability to execute an arbitrary command in a pod by opening the app with a specifically crafted address, but the PR description doesn't mention any considerations regarding the security of this feature.

In Connect, we had support for something like this (but with ssh servers), but we since removed it (see https://github.com/gravitational/teleport-private/issues/176, though the vulnerability persists).

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx (hunk 2 above)
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx:73]: Yeah, `DocumentSsh` was broken, I've fixed it (PR fixing it: https://github.com/gravitational/teleport/pull/41877 ) and followed same pattern, it also renders basically same 🤷

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx (hunk 3 above)
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.story.tsx:79]: Forgot to use in the `DocumentKubeExecWrapper` 😅 , fixed https://github.com/gravitational/teleport/pull/41466/commits/27fc01c91e7f6487e715dce10e0bdcb9da05b306

File: web/packages/teleport/src/Console/consoleContext.tsx [hunk 7]
```diff
@@ -186,28 +230,57 @@ export default class ConsoleContext {
...
+          command,
+        };
+        break;
+    }
 
-    const ttyUrl = cfg.api.ttyWsAddr
+    const ttyUrl = baseUrl
       .replace(':fqdn', getHostName())
       .replace(':clusterId', clusterId)
+      // .replace(':clusterName', serverId)
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/consoleContext.tsx]: Yep, removed.

File: web/packages/teleport/src/Console/DocumentKubeExec/useKubeExecSession.tsx
[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/useKubeExecSession.tsx:1]: Yes, but I also didn't want to couple the two too much and opted for some repeated code. The reason is that this is an MVP of a feature (general "kubectl in web UI") that is used to gauge interest and get feedback, there's a high chance it will be changed/replaced/removed soon. So I think it's better to keep them a bit apart, since we know we have plans to change it. The end goal feature might be also very similar to overall SSH experience and at that point we might combine the code for them.

**Comment by AntonAM** [comment]
@rosstimothy 

**UX:** 
I don't really see benefit of splitting input into two, `namespace/pod` seems like a natural format, also we have optional container one could specify with `namespace/pod/container`, which would require third input.

**Security:**
Ok, I'll look into splitting the flow 👍

**Comment by dmsergeevN26** [comment]
Amazing! Is it possible to add support for kubectl debug in addition/instead of exec? Debug is the preferred way to debug pods nowadays. Does this support moderation? I hope my questions are not out of place! Thank you!

**Comment by AntonAM** [comment]
@dmsergeevN26 We're currently evaluating paths to expand kubernetes access in Web UI, this is our first step in this direction. Debug functionality might be implemented in the future. You're feedback is welcomed!

File: web/packages/teleport/src/Console/consoleContext.tsx [hunk 8]
```diff
@@ -186,28 +230,57 @@ export default class ConsoleContext {
...
+        break;
+      case 'k8s':
+        const splits = login.split('/');
+        ttyParams = {
+          kubeCluster: resourceName,
+          namespace: splits[0],
+          pod: splits[1],
+          container: splits?.[2] || '',
+          isInteractive,
+          command,
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/consoleContext.tsx]: @ravicious I've split the flow as Tim suggested - now there's no just an url to click on, user always have to provide input (see recording). I've also removed snapshot tests and added regular test (didn't go as far as sending bytes over connection in there, it seems to be more complicated maybe will return to it in later PR, but it checks rendering)

https://github.com/gravitational/teleport/assets/964862/a1373a4d-1066-4754-aa2e-1b4b61b309f0

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.tsx [hunk 9]
```diff
@@ -67,7 +71,24 @@ export default function DocumentKubeExec({ doc, visible }: Props) {
           errorText={webauthn.errorText}
         />
       )}
-      {status === 'initialized' && terminal}
+
+      {status === 'initialized' && !sentData && (
```
Comment (ravicious) [inline web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.tsx:75]: I feel like it might be a good idea to introduce an explicit status instead of using an auxiliary boolean variable.

`DocumentSsh` has two statuses, `initialized` and `loading`. `DocumentKubeExec` at the moment has three, `initialized`, `loading`, `disconnected`.

What do you think about introducing a new status called `waiting-for-exec-data`? This way where `useKubeExecSession` now calls `setStatus('initialized')`, it'd call `setStatus('waiting-for-exec-data`)` instead.

Instead of passing an arbitrary `onExec` callback to `KubeExecData`, `useKubeExecSession` could return a function called `sendKubeExecData` that calls the corresponding method on `tty` and switches the status to `initialized`.

Long term I think it'd be easier for someone new to understand how different statuses map to actual behavior of the component, as it'd be a matter of understanding four statuses vs three + how `sentData` interacts with them.

File: web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+                >
+                  Close
+                </ButtonSecondary>
+                <ButtonPrimary
+                  type="submit"
+                  width="45%"
+                  size="large"
+                  onClick={e => {
+                    e.preventDefault();
+                    validator.validate() && podExec();
```
Comment (ravicious) [inline web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx:151]: I was like "How does this work without `<form>` having `onSubmit` on it?". I looked into the spec and it turns out that indeed, just having a `<button type="submit">` with an `onClick` is enough, TIL.

https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#implicit-submission

**Review by ravicious — APPROVED** [review]
Huge, it seems much better now in terms of security. Thanks for adding real tests instead of snapshot tests!

File: lib/web/apiserver.go [hunk 11]
```diff
@@ -3376,16 +3382,28 @@ func (h *Handler) podConnect(
...
+	if err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, net.ErrClosed) || terminal.IsOKWebsocketCloseError(trace.Unwrap(err)) {
+			return nil, nil
+		}
+		var netError net.Error
+		if errors.As(trace.Unwrap(err), &netError) && netError.Timeout() {
+			return nil, trace.BadParameter("timed out waiting for pod exec request data on websocket connection")
+		}
+
+		return nil, err
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply lib/web/apiserver.go]: Handler's `WithClusterAuthWebSocket`, which podConnect uses, writes returned error to the web socket.

File: lib/web/apiserver.go [hunk 12]
```diff
@@ -3459,6 +3478,39 @@ func (h *Handler) podConnect(
...
+
+	messageType, bytes, err := ws.ReadMessage()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = ws.SetReadDeadline(time.Time{})
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to set read deadline for websocket connection")
+	}
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply lib/web/apiserver.go]: We don't enforce any read deadline for kube exec code. Also, we have ping loop going on, so if anything happens and pings stop it will close the connection?

File: lib/web/apiserver.go [hunk 13]
```diff
@@ -3402,7 +3420,7 @@ func (h *Handler) podConnect(
 
 	sess := session.Session{
 		Kind:                  types.KubernetesSessionKind,
-		Login:                 sctx.GetUser(),
+		Login:                 strings.Join([]string{execReq.Namespace, execReq.Pod, execReq.Container}, "/"),
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply lib/web/apiserver.go]: It was leftover from the hack, but for kube sessions login doesn't really matter, it just needs to be something, so we always set it to "root". I've changed it to be "root".

File: lib/defaults/defaults.go [hunk 14]
```diff
@@ -280,6 +280,10 @@ const (
 	// HeadlessLoginTimeout is how long to wait for user to approve/reject headless login request.
 	HeadlessLoginTimeout = SSOCallbackTimeout
 
+	// KubeExecDataWaitTimeout is how long server would wait for user to send pod exec data (namespace, pod name etc)
+	// on websocket connection, after user initiated the exec into pod flow.
+	KubeExecDataWaitTimeout = HeadlessLoginTimeout
+
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply lib/defaults/defaults.go]: Done https://github.com/gravitational/teleport/pull/41466/commits/0e41e2fbc63bb1979709392db0b9f5e5ab4b52f7

File: web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.tsx (hunk 9 above)
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/DocumentKubeExec.tsx]: Yep, it's cleaner that way and we get rid of "sentData", I've changed it 👍 https://github.com/gravitational/teleport/pull/41466/commits/32131809293420e45a31c3ee88cd50d22e6173ec

File: lib/web/apiserver.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (ravicious) [reply lib/web/apiserver.go]: FWIW, I had the same experience, but I assumed it's expected. An error message there would be nice though.

**Review by rosstimothy — COMMENTED** [review]
The new session creation functionality that exists within the terminal tab is still wired up for SSH servers:

https://github.com/gravitational/teleport/assets/39066650/37176c4e-adb6-431f-a5db-98e9b1194134

File: lib/web/apiserver.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (AntonAM) [reply lib/web/apiserver.go]: Turned out that even though `WithClusterAuthWebSocket` was writing an error to the websocket, it was doing that incorrectly, missing version field
image

But fixing that doesn't help, because client side tty doesn't actually know what to do with an error, it doesn't have a handler for that type of message, so I guess that never worked - it also doesn't work for the ssh nodes connections, if you return an error in the handler, it also just displays "disconnected"

image

We can fix it, but since it's a general problem I would like it to be a separate PR (I'll open an issue).

File: web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+                    placeholder="/bin/bash"
+                    label="Command to execute"
+                    width="66%"
+                    onChange={e => setExecCommand(e.target.value)}
+                    toolTipContent={
+                      <Text>
+                        The command that will be executed inside the target pod.
+                      </Text>
+                    }
+                  />
```
[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx:116]: Yeah, that's a nodes list page, so currently it's only about SSH. I guess it would be part of expanding kubectl in webui work to add kube there, along with the possibility of joining session in the web ui.

Regarding length limit, it looks like we generally don't do that in the UI, we don't have length rule in our standard validation rule set. Kubernetes has limits on length of namespace/pod/container so we could enforce it, but I think we can just do that on the backend, returning something like `trace.BadParameter("Namespace is too long, should be less than 253 symbols")`, I see it as a very rare edge case and nobody with good intentions probably would hit it.

**Comment by AntonAM** [comment]
@rudream friendly ping

File: web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx:116]: Yeah, I meant that we only need to enforce the limit on the backend without fiddling with validation for that on the client side, since it can't be an honest mistake.

[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx:116]: Added validation on the backend: https://github.com/gravitational/teleport/pull/41466/commits/4d8974017f65e6c3588fb4b8d77ba760b1b9a2d3

**Review by rudream — APPROVED** [review]
Frontend changes LGTM

File: lib/web/apiserver.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (AntonAM) [reply lib/web/apiserver.go]: Opened: https://github.com/gravitational/teleport/issues/42626

File: web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx:116]: We can not close the document on graceful disconnect, then this UI won't show up.

[the remark this answers is not in the document]
Comment (AntonAM) [reply web/packages/teleport/src/Console/DocumentKubeExec/KubeExecDataDialog.tsx:116]: Yeah, but that UI is general, I mean, I think it should be there - you might want to open an ssh connection alongside the kube exec connection. But I think it should be ok for the first version? And once we expand UI support for kube in web, it will be more natural, where you will be able to choose which type of resource you want to connect to.

**Comment by roraback** [comment]
> @roraback do you have opinions about the design/UX in question?

I'm okay with a split. My bigger questions would be: 
1. Is the CLI flow first because we expect it to be more heavily used than the browser flow?
2. How might we make the decision between CLI and web browser a bit clearer? Right now it feels like you have CLI instructions (without ever explicitly saying it's about the CLI) and an add-on that says OR you can do this other thing.
3. Can we make the "Step 1" instruction more explicitly about the CLI? (though this question might become moot if we make Q2 more clearly a decision between using a CLI app and executing a command through the browser.)

**Comment by AntonAM** [comment]
I've added additional text, making it more clear that first option is about CLI, so I think it is now more distinct and clear that first option is about cli and second option is about Web UI. And as @rosstimothy mentioned it was decided to put exec down there to not emphasize it.

image
